### ISSUE TITLE
Playing alot of small sounds stops after some time

### ISSUE THREAD

**Issue by LocutusOfPenguin** [spatialaudio/python-sounddevice#51]
When i play lots of ~0.5sec voice files, the system stops working.
In general, i get no error message just the loudspeaker keep silent.

But right now i get this (and i use this to ask for help):
sounddevice.PortAudioError: Invalid stream pointer

The code im using can be seen here (but its similar to the "play_file" example):
https://github.com/LocutusOfPenguin/picochess/blob/master/talker/picotalker.py

As said, this effect is after some "moves" , and each move is around 1-3 small voice files.

**Comment by mgeier** [comment]
Can you please give step-by-step instructions how to reproduce this?
Or even better, could you create a stand-alone runnable example that produces the error?
You could e.g. create a Gist including the necessary sound files.

**Comment by LocutusOfPenguin** [comment]
its abit hard for me..
I tried to write a script sending lots of my sounds to the soundsystem..And y can guess, that is working.

But....if y checkout picochess , and start it with
"picochess -c -uvoice en:al -cvoice de:janniclas" and play a few moves y can come to the error..
=> its so often, y wont miss it...but its not always at same place.

I changed my code to be absolute sure, the code is not called by two threads at same time (cause i thinked that was the cause), but it doesnt helped to fix this problem.
I saw when two threads calling "play" at same time, iresulting in "keep silent" afterwards

If y dont /cant do "checkout picochess" and do enter moves, perhaps y give me a hint, how to find out the cause (=narrow the search) - see above.
Jürgen

**Comment by mgeier** [comment]
Yes, in general you should be very careful with multiple threads. [This page about PortAudio](https://app.assembla.com/spaces/portaudio/wiki/Tips_Threading) says:

> At the moment, the thread safety of PortAudio is technically unspecified and hence platform/implementation dependent

It is surely the safest bet just to avoid multiple threads accessing PortAudio.

I am willing to try to install your stuff, but only if you give me step-by-step instructions. I don't have a RaspPi, BTW.

**Comment by LocutusOfPenguin** [comment]
You dont need a rasp (i have this stopped voice at my ubuntu too)..and at best, y dont need a dgt chess board - just in this case need to type in moves on console (which is not too user friendly).

But y need to play chess...can y? if not , y might try to enter "lever:w" instead of yr move (switching sides) ...but this only using one (out of two) voices.


git clone https://github.com/LocutusOfPenguin/picochess.git
install a virtual env (not really needed - but keep yr system clean)
pip3 install --upgrade -r requirements.txt

then y should have a working picochess system NOW:

python3 picochess.py -c -uvoice en:dgt -cvoice en:al -l debug -lf picochess.log

starts a picochess in console mode.

moves are typed in like "e2e4" or "e1g1" for 0-0 and after each move need to give "go" ...same if y get a move from chess engine (type "go"). This part is only useful, if y know the chess rules ;-)

If y dont know chess y can only type "lever:l" & "go" and again and again ...(always switch sides)

$$$$
I see now, it even only speaks the first part of the move , say "n","f","3" only "n" spoken ...hmmm

**Comment by mgeier** [comment]
OK, I tried it:

```
$ python3 picochess.py -c -uvoice en:dgt -cvoice en:al -l debug -lf picochess.log
Traceback (most recent call last):
  File "picochess.py", line 888, in <module>
    main()
  File "picochess.py", line 475, in main
    args.engine = el[0]['file']  # read the first engine filename and use it as standard
IndexError: list index out of range
```

**Comment by LocutusOfPenguin** [comment]
pls run
./build_engines.py

**Comment by mgeier** [comment]
no change, same error

**Comment by LocutusOfPenguin** [comment]
hmmm....what platform (x86_64?) y running on? 
y can try to see on engines.py  @ read_engine_ini what yr system is doing.

another way is to give the engines path with "-e" inside...but i think, yr path are not ok (right now)

**Comment by mgeier** [comment]
Yes, x84_64 Linux (Debian testing), Python 3.5.2

If you want me to try this, you'll have to tell me what to do.

If the `play()` function doesn't work for you, you could try writing your own callback function. This way you could avoid stopping and starting the PortAudio stream over and over again.

**Comment by LocutusOfPenguin** [comment]
pull again
copy the dgt/test/x86_64/* inside yr engines/x86_64
try again

if y have "ogg123" installed ...y must deactivate it in talker/picotalker.py , otherwise it not use yr system

**Comment by mgeier** [comment]
I've changed my strategy because it's too confusing to test this within such a complicated application.

I just tried this:

```python
for _ in range(150):
    sd.play(some_array, 44100)
```

Every once in a while this creates the "invalid stream pointer" error you mentioned.
Sometimes it just freezes, which is probably similar to the case where you hear no output and see no errors (because probably one of your threads freezes).

So I guess I found a way to reproduce your errors, now let's try to find and solve the problem behind them!

**Comment by LocutusOfPenguin** [comment]
Great.
Taking picochess is indeed lot of code (irrelevant to this issue).
Hope, y find the cause.

**Comment by mgeier** [comment]
@LocutusOfPenguin Can you please try if #54 fixes your problem?

**Comment by mgeier** [comment]
@LocutusOfPenguin Any news?

**Comment by mgeier** [comment]
@LocutusOfPenguin Did you have a chance to test #54?

**Comment by mgeier** [comment]
I've just merged #45 which hopefully closes this issue.

If not, please re-open it!
